File: examples/RangeIteratorT.js
#!/usr/bin/env node

/* global
    before,
    describe,
    it
*/

/* jshint
    esversion: 6
*/

// -----------------
// RangeIteratorT.js
// -----------------

"use strict";

const assert   = require('assert');
const _        = require('lodash');
const readline = require('readline');

class range_iterator_1 {
    constructor (b, e) {
        this.b = b;
        this.e = e;}

    [Symbol.iterator] () {
        return this;}

    next () {
        let n = {
            value: this.b,
            done:  (this.b == this.e)};
        ++this.b;
        return n;}}

function* range_iterator_2 (b, e) {
    while (b != e) {
        yield b;
        ++b;}}

describe('RangeIterator',
    function () {
        let a;

        before(function () {
            a = [
                (b, e) => {
                    return new range_iterator_1(b, e);},
                range_iterator_2,
                (b, e) => {
                    const x = _.range(b, e);
                    return x[Symbol.iterator]();}];});

        it('test1',
            function () {
                for (let f of a) {
                    const p = f(2, 2);
                    assert(p === p[Symbol.iterator]());
                    const n = p.next();
                    assert(n.done);}});

        it('test2',
            function () {
                for (let f of a) {
                    const p = f(2, 3);
                    assert(p === p[Symbol.iterator]());
                    let n = p.next();
                    assert(!n.done);
                    assert(n.value === 2);
                    n = p.next();
                    assert(n.done);}});

        it('test3',
            function () {
                for (let f of a) {
                    const p = f(2, 4);
                    assert(p === p[Symbol.iterator]());
                    let n = p.next();
                    assert(!n.done);
                    assert(n.value === 2);
                    n = p.next();
                    assert(!n.done);
                    assert(n.value === 3);
                    n = p.next();
                    assert(n.done);}});

        it('test4',
            function () {
                for (let f of a) {
                    const p = f(2, 5);
                    assert(p === p[Symbol.iterator]());
                    const a = [];
                    for (let v of p)
                        a.push(v);
                    assert(_.isEqual(a, [2, 3, 4]));}});});

File: examples/RangeIteratorT.py
#!/usr/bin/env python3

# pylint: disable = bad-whitespace
# pylint: disable = invalid-name
# pylint: disable = missing-docstring
# pylint: disable = too-few-public-methods

# -----------------
# RangeIteratorT.py
# -----------------

from typing   import Iterator
from unittest import main, TestCase

class range_iterator_1 (Iterator[int]) :
    def __init__ (self, b: int, e: int) -> None :
        self.b = b
        self.e = e

    def __iter__ (self) -> Iterator[int] :
        return self

    def __next__ (self) -> int :
        if self.b == self.e :
            raise StopIteration()
        v = self.b
        self.b += 1
        return v

class MyUnitTests (TestCase) :
    def setUp (self) :
        self.a = [
            range_iterator_1,
            lambda b, e : iter(range(b, e))]

    def test_1 (self) :
        for f in self.a :
            with self.subTest(msg=f.__name__) :
                p = f(2, 2)
                self.assertIs(p, iter(p))
                self.assertRaises(StopIteration, next, p)

    def test_2 (self) :
        for f in self.a :
            with self.subTest(msg=f.__name__) :
                p = f(2, 3)
                self.assertIs(p, iter(p))
                self.assertEqual(next(p), 2)
                self.assertRaises(StopIteration, next, p)

    def test_3 (self) :
        for f in self.a :
            with self.subTest(msg=f.__name__) :
                p = f(2, 4)
                self.assertIs(p, iter(p))
                self.assertEqual(next(p), 2)
                self.assertEqual(next(p), 3)
                self.assertRaises(StopIteration, next, p)

    def test_4 (self) :
        for f in self.a :
            with self.subTest(msg=f.__name__) :
                p = f(2, 5)
                self.assertIs(p, iter(p))
                a = []
                for v in p :
                    a.append(v)
                self.assertEqual(a, [2, 3, 4])

    def test_5 (self) :
        for f in self.a :
            with self.subTest(msg=f.__name__) :
                p = f(2, 5)
                self.assertIs(p, iter(p))
                self.assertEqual(list(p), [2, 3, 4])
                self.assertEqual(list(p), [])

if __name__ == "__main__" : # pragma: no cover
    main()
